proposals/reference-types/Overview.md

```diff
@@ -0,0 +1,204 @@
+# Reference Types for WebAssembly
+
+TODO: more text, motivation, explanation
+
+## Introduction
+
+Motivation:
+
+* Easier and more efficient interop with host environment
+  - allow host references to be represented directly by type `anyref`
+  - without having to go through tables, allocating slots, and maintaining index bijections at the boundaries
+
+* Basic manipulation of tables inside Wasm
+  - allow representing data structures containing references
+by repurposing tables as a general memory for opaque data types
+  - allow manipulating function tables from within Wasm.
+
+* Set the stage for later additions:
+
+  - Typed function references (see below)
+  - Exception references (see exception handling proposal)
+  - A smoother transition path to GC (see GC proposal)
+
+Get the most important parts soon!
+
+Summary:
+
+* Add a new type `anyref` that can be used as both a value type and a table element type.
+
+* Also allow `anyfunc` as a value type.
+
+* Introduce instructions to get and set table slots.
+
+* Allow multiple tables.
+
+Notes:
+
+* This extension does not imply GC by itself, only if host refs are GCed pointers!
+
+* Reference types are *opaque*, i.e., their value is abstract and they cannot be stored into linear memory. Tables are used as the equivalent.
+
+
+## Language Extensions
+
+Typing extensions:
+
+* Introduce `anyref`, `anyfunc`, and `nullref` as a new class of *reference types*.
+  - `reftype ::= anyref | anyfunc | nullref`
+
+* Value types (of locals, globals, function parameters and results) can now be either numeric types or reference types.
+  - `numtype ::= i32 | i64 | f32 | f64`
+  - `valtype ::= <numtype> | <reftype>`
+  - locals with reference type are initialised with `null`
+
+* Element types (of tables) are equated with reference types.
+  - `elemtype ::= <reftype>`
+
+* Introduce a simple subtype relation between reference types.
+  - `nullref < t` for all reftypes `t`
+  - `t < anyref` for all reftypes `t`
+  - `anyfunc < anyref`
+
+
+New/extended instructions:
+
+* The new instruction `ref.null` evaluates to the null reference constant.
+  - `ref.null : [] -> [nullref]`
+
+* The new instructions `table.get` and `table.set` access tables.
+  - `table.get $x : [i32] -> [t]` iff `t` is the element type of table `$x`
+  - `table.set $x : [i32 t] -> []` iff `t` is the element type of table `$x`
+  - `table.fill $x : [i32 i32 t] -> []` iff `t` is the element type of table `$x`
+
+* The `call_indirect` instruction takes an additional table index as immediate that identifies the table it calls through.
```

Comment:
"additional" sounds like a breaking change; can you nuance to say "the current restriction that the table index immediate must be 0 is removed"?

Reply:
Removed "additional", but note that it is new in both the abstract syntax and the text format.